### ISSUE TITLE
Smeared imagery on some mobile devices

### ISSUE THREAD

**Issue by kring** [AnalyticalGraphicsInc/cesium#2541]
Originally reported here:
https://groups.google.com/d/topic/cesium-dev/ouQyJY0LiRs/discussion

The immediate cause is that the `webMercatorT` vertex attribute is receiving incorrect values in the imagery reprojection shader.  The attribute is set with the following code (in `ImageryLayer`):

```
reproject.vertexArray.getAttribute(1).vertexBuffer.copyFromArrayView(webMercatorT);
```

I've confirmed that the `webMerctorT` `Float32Array` is correct.  On the problematic mobile devices (HTC One M7 in my case), this line actually corrupts the attribute - values that were provided when the vertex array was originally created are actually replaced with garbage.

The best part is that this problem goes away if we don't use `OES_vertex_array_object`.  Comment out `this._vertexArrayObject = getExtension(gl, ['OES_vertex_array_object']);` in `Context` and everything is fine.

I tried to demonstrate this problem with a simpler test, but apparently my test is too simple because it passed even on my M7.  Here's what I tried:

```
    it('can update vertex buffer', function() {
        var vs = '\
            attribute vec4 position;\
            attribute float test;\
            varying float v_test;\
            void main()\
            {\
                v_test = test;\
                gl_PointSize = 1.0;\
                gl_Position = position;\
            }';
        var fs = '\
            varying float v_test;\
            void main()\
            {\
                gl_FragColor = vec4(v_test == 0.5);\
            }';

        var context = createContext();
        var sp = context.createShaderProgram(vs, fs);

        var va = context.createVertexArray([
            {
                index : sp.vertexAttributes.position.index,
                vertexBuffer : context.createVertexBuffer(new Float32Array([0, 0, 0, 1, 0, 0, 0, 1]), BufferUsage.STATIC_DRAW),
                componentsPerAttribute : 4
            },
            {
                index : sp.vertexAttributes.test.index,
                vertexBuffer : context.createVertexBuffer(new Float32Array([0.0, 0.0]), BufferUsage.STATIC_DRAW),
                componentsPerAttribute : 1
            }
        ]);

        ClearCommand.ALL.execute(context);
        expect(context.readPixels()).toEqual([0, 0, 0, 0]);

        var command = new DrawCommand({
            primitiveType : PrimitiveType.POINTS,
            shaderProgram : sp,
            vertexArray : va
        });

        va.getAttribute(1).vertexBuffer.copyFromArrayView(new Float32Array([0.5, 0.5]));

        command.execute(context);
        expect(context.readPixels()).toEqual([255, 255, 255, 255]);

        sp = sp.destroy();
        va = va.destroy();
    });
```

Maybe @pjcozzi or @bagnell could take a quick look at the reprojection code in `ImageryLayer` to see if I'm doing something stupid, maybe along the lines of that impossible-to-debug Firefox-on-Mac-only problem from awhile back that turned out to be my failure to explicitly specify vertex attribute indices.

If not, I guess we'll need to try harder to turn this into a test case to report it as a bug to the browser vendors.  Since it happens in both Chrome and Firefox on this device, though, it's probably a GPU/driver bug and we'll need a workaround instead (such as not using VAOs on these devices).

**Comment by pjcozzi** [comment]
Try storing the vertex buffer in a separate variable for accessing it, which is commonly done through the primitives.

Perhaps there is a bug in `VertexArray` where `_attributes` is mis-ordered when VAOs are present.

Perhaps there is a state issue since that test didn't fail...I wouldn't think so though.

**Comment by kring** [comment]
I'm still a long way from an answer, but here are some interesting new facts:
* Creating a new vertex array for every reprojection fixes the problem.
* Saving off the vertex buffer immediately after creating that array and using that in later frames doesn't have any effect (still broken).
* Combining the two vertex buffers into a single vertex buffer, so the one and only vertex buffer is updated on each reprojection, does NOT fix the problem.

I wasn't able to identify a Sandcastle example (or any other example) that causes `Primitive` to invoke this problematic `copyFromArrayView`.  `BillboardCollection` uses it via `VertexArrayFacade`, but code that exercises that (such as simple.czml) works fine on this device.

So I'll keep looking at it when I get a chance, but ideas are welcome!

**Comment by pjcozzi** [comment]
> Creating a new vertex array for every reprojection fixes the problem.

The vertex buffers are being created with `BufferUsage.STATIC_DRAW`, which is just a hint to the driver, but it is wrong in this case.  We want `BufferUsage.STREAM_DRAW` since we rewrite the buffer each draw call (like streaming video).

In my experience, some drivers ignore this hint; on others, it has a significant performance impact.  I have ran into cases where I said something was `STATIC`, and then wrote to it and it was either really slow or didn't work (I can't remember) so I could expect a bug here.

Also, have you tested Chrome and Firefox?  Stable and Canary?

**Comment by kring** [comment]
> The vertex buffers are being created with BufferUsage.STATIC_DRAW

No, there are two vertex buffers, one for positions and one for the texture coordinates.  The positions never change and are created STATIC_DRAW.  The texture coordinate are updated every time and are created STREAM_DRAW.
https://github.com/AnalyticalGraphicsInc/cesium/blob/master/Source/Scene/ImageryLayer.js#L835

Chrome and Firefox both fail.  I haven't tried beta versions, but I will next time Kristen isn't using her phone. :)

**Comment by pjcozzi** [comment]
Ah, I was only looking at the example above.  Maybe try changing them both to `STREAM` or `DYNAMIC` and also updating the positions each frame.  It is possible the driver has a coalescing bug.

Are you running the latest OS?  How prominent is this bug?  It would be a shame to have to put a hacky workaround in.

**Comment by kring** [comment]
It's the latest OS available on this phone currently (though apparently it's supposed to get an update to 5.0 tomorrow).  I might hold off on updating, though, for fear of losing the ability to reproduce the problem.  Even if 5.0 fixes it, very few devices have Lollipop yet.  Someone on the forum reported a whole long list of devices it happens on, and when it happens Cesium is totally unusable.  So I think it's worth a workaround.

Will play with it more when I get a chance, but that probably won't be until Friday (I'm going to a conference starting tomorrow morning).

**Comment by ianwalberg** [comment]
Any update or a workaround for this issue?

Thanks, Ian

**Comment by kring** [comment]
No, update, but I mentioned a workaround in my original comments in this issue:

> Comment out `this._vertexArrayObject = getExtension(gl, ['OES_vertex_array_object']);` in `Context` and everything is fine.

By `Context` I meant `Source/Renderer/Context.js`.  The [Contributor's Guide](https://github.com/AnalyticalGraphicsInc/cesium/wiki/Contributor%27s-Guide) has information on building your own version of Cesium including this change (it's not too hard).

**Comment by ianwalberg** [comment]
Kevin,

Thanks we will try the work around this week.

Regards

Ian

**Comment by hpinkos** [comment]
Also reported here: https://groups.google.com/forum/?hl=en#!topic/cesium-dev/4b1COszHwCE

![image](https://cloud.githubusercontent.com/assets/3451886/19358945/0106e76a-9146-11e6-99de-5275f492e8d4.png)

**Comment by kring** [comment]
I don't have a device I can test this on anymore, but it's very likely that #4339 will fix it.

**Comment by kring** [comment]
Actually sorry #4339 won't fix it completely, because it still does reprojection for images in terrain tiles that cross 85 degrees latitude.  But it should make the rest of the globe work at least.
